Sources/GRPCHTTP2Core/Client/Connection/Connection.swift

```diff
@@ -0,0 +1,434 @@
+        channel.eventLoop.makeCompletedFuture {
+          let streamHandler = GRPCClientStreamHandler(
+            methodDescriptor: descriptor,
+            scheme: scheme,
+            outboundEncoding: compression,
+            acceptedEncodings: self.enabledCompression,
+            maximumPayloadSize: maxRequestSize
+          )
+          try channel.pipeline.syncOperations.addHandler(streamHandler)
+
+          return try NIOAsyncChannel(
+            wrappingChannelSynchronously: channel,
+            configuration: NIOAsyncChannel.Configuration(
+              isOutboundHalfClosureEnabled: true,
+              inboundType: RPCResponsePart.self,
+              outboundType: RPCRequestPart.self
+            )
+          )
+        }
+      }
+
+      return Stream(wrapping: stream, descriptor: descriptor)
+    } catch {
+      throw RPCError(code: .unavailable, message: "subchannel is unavailable", cause: error)
+    }
+  }
+
+  private func consumeConnectionEvents(
+    _ connectionEvents: NIOAsyncChannelInboundStream<ClientConnectionEvent>
+  ) async {
+    do {
+      var channelCloseReason: ClientConnectionEvent.CloseReason?
+
+      for try await connectionEvent in connectionEvents {
+        switch connectionEvent {
+        case .closing(let reason):
+          self.state.withLockedValue { $0.closing() }
+
+          switch reason {
+          case .goAway(let errorCode, let reason):
+            // The connection will close at some point soon, yield a notification for this
+            // because the close might not be imminent and this could result in address resolution.
+            self.event.continuation.yield(.goingAway(errorCode, reason))
+          case .idle, .keepaliveExpired, .initiatedLocally:
+            // The connection will be closed imminently in these cases there's no need to do
+            // anything.
+            ()
+          }
+
+          // Take the reason with the highest precedence. A GOAWAY may be superseded by user
+          // closing, for example.
+          if channelCloseReason.map({ reason.precedence > $0.precedence }) ?? true {
+            channelCloseReason = reason
+          }
+        }
+      }
+
+      let connectionCloseReason: Self.CloseReason
+      switch channelCloseReason {
+      case .keepaliveExpired:
+        connectionCloseReason = .keepaliveTimeout
+
+      case .idle:
+        // Connection became idle, that's fine.
+        connectionCloseReason = .idleTimeout
+
+      case .goAway:
+        // Remote peer told us to GOAWAY.
+        connectionCloseReason = .remote
+
+      case .initiatedLocally, .none:
+        // Shutdown was initiated locally.
+        connectionCloseReason = .initiatedLocally
+      }
+
+      // The connection events sequence has finished: the connection is now closed.
+      self.state.withLockedValue { $0.closed() }
+      self.finishStreams(withEvent: .closed(connectionCloseReason))
```

Comment:
Yes: we only do this because the connection events sequence has finished meaning that the connection is now closed (meaning the streams must also be closed). Note that when we receive the GOAWAY in the for-await loop we don't close the connection.